Dependencies/xr/Source/OpenXR/XR.cpp

```diff
@@ -637,12 +686,152 @@ namespace xr
+                actionInfo.subactionPaths = ActionResources.ControllerSubactionPaths.data();
+                XrCheck(xrCreateAction(ActionResources.ActionSet, &actionInfo, &ActionResources.ControllerGetThumbstickAxesAction));
+                // For each controller subaction
+                for (size_t idx = 0; idx < ActionResources.CONTROLLER_SUBACTION_PATH_PREFIXES.size(); ++idx)
+                {
+                    // Create suggested binding
+                    std::string path{ ActionResources.CONTROLLER_SUBACTION_PATH_PREFIXES[idx] };
+                    path.append(ActionResources.CONTROLLER_GET_THUMBSTICK_AXES_PATH_SUFFIX);
+                    bindings.push_back({ ActionResources.ControllerGetThumbstickAxesAction });
+                    XrCheck(xrStringToPath(instance, path.data(), &bindings.back().binding));
+                }
+            }
+
+            // Create controller get thumbstick click action and suggested bindings
+            {
+                XrActionCreateInfo actionInfo{ XR_TYPE_ACTION_CREATE_INFO };
+                actionInfo.actionType = XR_ACTION_TYPE_BOOLEAN_INPUT;
+                strcpy_s(actionInfo.actionName, ActionResources.CONTROLLER_GET_THUMBSTICK_CLICK_ACTION_NAME);
+                strcpy_s(actionInfo.localizedActionName, ActionResources.CONTROLLER_GET_THUMBSTICK_CLICK_ACTION_LOCALIZED_NAME);
+                actionInfo.countSubactionPaths = static_cast<uint32_t>(ActionResources.ControllerSubactionPaths.size());
+                actionInfo.subactionPaths = ActionResources.ControllerSubactionPaths.data();
+                XrCheck(xrCreateAction(ActionResources.ActionSet, &actionInfo, &ActionResources.ControllerGetThumbstickClickAction));
+                // For each controller subaction
+                for (size_t idx = 0; idx < ActionResources.CONTROLLER_SUBACTION_PATH_PREFIXES.size(); ++idx)
+                {
+                    // Create suggested binding
+                    std::string path{ ActionResources.CONTROLLER_SUBACTION_PATH_PREFIXES[idx] };
+                    path.append(ActionResources.CONTROLLER_GET_THUMBSTICK_CLICK_PATH_SUFFIX);
+                    bindings.push_back({ ActionResources.ControllerGetThumbstickClickAction });
+                    XrCheck(xrStringToPath(instance, path.data(), &bindings.back().binding));
+                }
+            }
+
+            // Provide default Microsoft suggested binding to instance
+            XrInteractionProfileSuggestedBinding microsoftSuggestedBindings{ XR_TYPE_INTERACTION_PROFILE_SUGGESTED_BINDING };
+            XrCheck(xrStringToPath(instance, ActionResources.MICROSOFT_XR_INTERACTION_PROFILE, &microsoftSuggestedBindings.interactionProfile));
+            microsoftSuggestedBindings.suggestedBindings = bindings.data();
+            microsoftSuggestedBindings.countSuggestedBindings = (uint32_t)bindings.size();
+            XrCheck(xrSuggestInteractionProfileBindings(instance, &microsoftSuggestedBindings));
```

Comment:
As above, unless we're consciously in a place that is designated (and named) as Microsoft-specific, we should make sure that we're not doing things that will clash with the broader vision of OpenXR. Especially with Oculus and Valve now getting OpenXR runtimes into the public, we should anticipate people wanting to add support for those in the near future and try to plan ways for the implementations to share what can be shared and clearly designate what must be separated.

Please correct me if I'm misreading, but it looks like parts of this PR are _replacing_ the KHR simple controller as default with a Microsoft-specific alternative. If that's what's intended, I recommend reconsidering. The KHR simple controller exists as a default partly because it encapsulates a minimum functionality set that many different vendor implementations can map to. I think it would be better for Microsoft's more custom and feature-rich implementation to live alongside the simplistic default as opposed to replacing it. This will both leave the simplistic default intact and help establish the patterns that will eventually allow multiple feature-rich controller implementations (Oculus, Vive, Index, etc.) to coexist in the code alongside what we're currently actively implementing.

Reply:
thanks for bringing this up Justin, initially I thought there will be a runtime way of determining what controller is being used in openXR  but did not find any specific details of that sort. Later, after chatting with Tym, I understood that we would like to support all windows MR controllers as a first step. But this makes sense, I can dig deeper to see if it is possible to support a fall back option incase something is not a windows controller. Also, I would like to get @tymlipari 's perspective on this.